sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/ContinuousProgressReporter.scala

```diff
@@ -0,0 +1,182 @@
+      sink.toString,
+      sinkCommitProgress.map(_.numOutputRows))
+
+    val newProgress = new StreamingQueryProgress(
+      id = id,
+      runId = runId,
+      name = name,
+      timestamp = formatTimestamp(currentTriggerStartTimestamp),
+      batchId = epochId,
+      durationMs = new java.util.HashMap(
+        (currentDurationsMs(epochId)._2 ++ recordDurationMs).toMap.mapValues(long2Long).asJava),
+      eventTime = new java.util.HashMap(executionStats.eventTimeStats.asJava),
+      stateOperators = executionStats.stateOperators.toArray,
+      sources = sourceProgress.toArray,
+      sink = sinkProgress)
+
+    if (hasNewData) {
+      // Reset noDataEventTimestamp if we processed any data
+      lastNoDataProgressEventTime = Long.MinValue
+      updateProgress(newProgress)
+    } else {
+      val now = triggerClock.getTimeMillis()
+      if (now - noDataProgressEventInterval >= lastNoDataProgressEventTime) {
+        lastNoDataProgressEventTime = now
+        updateProgress(newProgress)
+      }
+    }
+
+    currentStatus = currentStatus.copy(isTriggerActive = false)
+  }
+
+  protected def extractSourceToNumInputRows(t: Option[Any] = None)
+      : Map[SparkDataStream, Long] = {
+    require(t.isDefined && t.get.isInstanceOf[EpochStats])
+    Map(sources(0) -> t.get.asInstanceOf[EpochStats].inputRows)
+  }
+
+  /**
+   * Record the offsets range this trigger will process. Call this before updating
+   * `committedOffsets` in `StreamExecution` to make sure that the correct range is recorded.
+   */
+  override protected def recordTriggerOffsets(
+      from: StreamProgress,
+      to: StreamProgress,
+      epochId: Long): Unit = {
+    checkQueueBoundaries()
+    currentTriggerStartOffsets.put(epochId, from.mapValues(_.json))
+    currentTriggerEndOffsets.put(epochId, to.mapValues(_.json))
+  }
+
+  /** Records the duration of running `body` for the next query progress update. */
+  def reportTimeTaken[T](triggerDetailKey: String, epochId: Long)(body: => T): T = {
+    checkQueueBoundaries()
+    val durations = currentDurationsMs.getOrElseUpdate(epochId,
+      (triggerClock.getTimeMillis(), new mutable.HashMap[String, Long]()))
+    addOrUpdateTime(triggerDetailKey, durations._2)(body)
+  }
+
+  /**
+   * Records the duration of running `body` for once time, which will not be cleared
+   * when start a new trigger.
+   */
+  protected def recordTimeTaken[T](triggerDetailKey: String)(body: => T): T = {
+    addOrUpdateTime(triggerDetailKey, recordDurationMs)(body)
+  }
+
+  private def checkQueueBoundaries(): Unit = {
+    if (currentDurationsMs.size > numProgressRetention) {
+      currentDurationsMs.remove(earliestEpochId)
+    }
+
+    if (currentTriggerStartOffsets.size > numProgressRetention) {
+      currentTriggerStartOffsets.remove(earliestEpochId)
+    }
+
+    if (currentTriggerEndOffsets.size > numProgressRetention) {
+      currentTriggerEndOffsets.remove(earliestEpochId)
+    }
+
+    earliestEpochId += 1
```

Comment:
Looks like, the `earliestEpochId ` will be also updated even no data removed from the metrics maps, this may cause some old data never got a chance to be removed.